docs/content/sips/016-inbound-websockets.md

```diff
@@ -0,0 +1,189 @@
+    // Returns a pollable which becomes ready when either the frame
+    // has been sent, or an error has occurred. When this pollable is
+    // ready, the `get` method will return `some`.
+    subscribe: func() -> pollable;
+    
+    // Returns the result of sending a frame to a client, once it
+    // has either completed successfully or errored.
+    //
+    // The outer `option` represents future readiness. Users can wait
+    // on this `option` to become `some` using the `subscribe` method.
+    //
+    // The outer `result` is used to retrieve the response or error
+    // at most once. It will be success on the first call in which the
+    // outer option is `some`, and error on subsequent calls.
+    //
+    // The inner `result` represents that either the frame was sent
+    // successfully, or that an error occurred.
+    //
+    // Note that a success result from this function does *not*
+    // guarantee that the client received the frame (or that it ever
+    // will). The frame will be sent asynchronously and might not
+    // be delivered due to an unexpected close event or network
+    // failure. Applications which require delivery confirmation
+    // must handle that themselves.
+    get: func() -> option<result<result<_, error-code>>>;
+}
+
+interface inbound-websocket-receive {
+  use wasi:http/types.{incoming-request, response-outparam};
+  use websocket-types.{frame, close-frame};
+
+  // Construct a new handler for the specified request, which the
+  // client has requested to be upgraded to a WebSocket.
+  //
+  // - `id`: a globally unique identifier which may be used to send
+  //   frames to this WebSocket using
+  //   `outgoing-websocket#send-frame`. Note that any application
+  //   may use this `id` to send frames to the WebSocket for as
+  //   long as it is connected, which means it should be treated as
+  //   a secret and shared only with trusted parties.
+  //
+  // - `request`: a resource providing the request's method, path,
+  //   headers, etc.
+  //
+  // - `response-out`: a resource which may be used to send a
+  //   response to the client. If (and only if) the response has a
+  //   status code of 101, then the connection will be upgraded to a
+  //   WebSocket.
+  handle-new: func(id: string, request: incoming-request, response-out: response-outparam);
+
+  // Handle the most recent frames received for the specified
+  // WebSockets. `frames` consists of a mapping from WebSocket ids
+  // to ordered lists of frames.
+  handle-frames: func(frames: list<tuple<string, list<received-frame>>);
+}
+
+interface inbound-websocket-send {
+  use websocket-types.{frame, future-send-frame};
+
+  // Attempt to send the specified frame to specified WebSocket.
+  //
+  // This will fail if the WebSocket is no longer open. Otherwise,
+  // if the frame is an instance of `close`, the connection will
+  // be closed after sending the frame.
+  send-frame: func(id: string, frame: frame) -> future-send-frame;
```

Comment:
Should this allow batched sends?

Reply:
I considered that, but I figured that if an app has multiple messages to send to a given client and wants to do that efficiently, they can bundle them into a single frame (i.e. handle it in the app-level protocol).  Also, we can always add new `send-frames` function later, if desired.  I don't have strong feelings about it, though.